test/controllers/products_controller_test.rb

```diff
@@ -79,32 +68,129 @@
+      it "redirects user" do
+        merchant = merchants(:merchant_1)
+        # do not do login
+        get edit_product_path(merchant.products.first.id)
 
-      must_respond_with :bad_request
-      expect(flash[:error]).must_include "Could not add new product"
-      expect(flash[:name]).must_include "can't be blank"
+        # must_respond_with :redirect
+        expect(flash[:error]).must_equal "You must be logged in to edit a product"
+      end
     end
   end
 
-  describe "logged out user" do
-    it "cannot create a product with valid data" do
-      # is setting merchant id to nil enough to test this?
-      new_product = { product: { name: "Something amazing", price: 1000 } }
-      expect {
-        post products_path, params: new_product
-      }.wont_change "Product.count"
+  describe "update" do
+    describe "logged in merchant" do
+      it "will update an existing product" do
+        merchant = perform_login
+
+        starter_input = {
+          name: "An intangible thing",
+          merchant_id: merchant.id,
+          description: "Its hard to hold onto",
+          photo_url: "a cute url",
+          stock: 100,
+          price: 999,
+        }
+
+        product_to_update = Product.create(starter_input)
+
+        test_input = {
+          "product": {
+            name: "A REALLY COOL intangible thing",
+            merchant_id: merchants(:merchant_1).id,
+            description: "Its hard to hold onto",
+            photo_url: "a cute url",
+            stock: 100,
+            price: 999,
+          },
+        }
+
+        expect {
+          patch product_path(product_to_update.id), params: test_input
+        }.wont_change "Product.count"
+
+        must_respond_with :redirect
+        product_to_update.reload
+        expect(product_to_update.name).must_equal test_input[:product][:name]
+        expect(product_to_update.merchant_id).must_equal test_input[:product][:merchant_id]
+        expect(product_to_update.description).must_equal test_input[:product][:description]
+        expect(flash[:success]).must_equal "Product updated successfully"
+      end
+
+      it "will not update an existing product if given bogus data" do
+        merchant = perform_login
+        product_to_update = products(:product_3)
+
+        test_input = {
+          "product": {
+            name: nil,
+            merchant_id: merchant.id,
+            description: "nice stuff wooowza",
+            photo_url: "https://yeahforsure.org",
+            stock: 100,
+            price: 999,
+          },
+        }
 
-      must_respond_with :bad_request
-      expect(flash[:error]).must_equal "You must be logged in to create a new product"
+        expect {
+          patch product_path(product_to_update.id), params: test_input
+        }.wont_change "Product.count"
+
+        must_respond_with :bad_request
+        # expect(flash[:name]).must_include "can't be blank" # not working
+      end
+    end
```

Comment:
this works fine. I uncommented and changed comment to working so you can still find it.

Reply:
thx! took out comment